src/Symfony/Component/HttpClient/RetryHttpClient.php

```diff
@@ -0,0 +1,100 @@
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpClient\Response\AsyncContext;
+use Symfony\Component\HttpClient\Response\AsyncResponse;
+use Symfony\Component\HttpClient\Retry\RetryBackOffInterface;
+use Symfony\Component\HttpClient\Retry\RetryDeciderInterface;
+use Symfony\Contracts\HttpClient\ChunkInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * Automatically retries the failing requests.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class RetryHttpClient implements HttpClientInterface
+{
+    use AsyncDecoratorTrait;
+
+    private $decider;
+    private $strategy;
+    private $maxRetries;
+    private $logger;
+
+    private $retryCount = 0;
+
+    /**
+     * @param int $maxRetries The maximum number of times to retry
+     */
+    public function __construct(HttpClientInterface $client, RetryDeciderInterface $decider, RetryBackOffInterface $strategy, int $maxRetries = 3, LoggerInterface $logger = null)
+    {
+        $this->client = $client;
+        $this->decider = $decider;
+        $this->strategy = $strategy;
+        $this->maxRetries = $maxRetries;
+        $this->logger = $logger ?: new NullLogger();
+    }
+
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options) {
+            $exception = null;
+            try {
+                // clone the chunk to let the original value of `$didthrow`
+                $clonedChunk = clone $chunk;
+                if ($clonedChunk->isTimeout() || null !== $clonedChunk->getInformationalStatus()) {
+                    yield $chunk;
+
+                    return;
+                }
+
+                // only retry first chunk
+                if (!$clonedChunk->isFirst()) {
+                    $context->passthru();
+                    yield $chunk;
+
+                    return;
+                }
+            } catch (TransportExceptionInterface $exception) {
+                // catch TransportExceptionInterface to send it to strategy.
+            }
+
+            if ($this->retryCount >= $this->maxRetries || !$this->decider->shouldRetry($method, $url, $options, $response = $context->getResponse(), $exception)) {
+                $context->passthru();
+                yield $chunk;
+
+                return;
+            }
+
+            $context->setInfo('retry', $this->retryCount);
+            $response->cancel();
+
+            $delay = $this->strategy->getDelay($this->retryCount, $method, $url, $options, $response = $context->getResponse(), $exception);
+            ++$this->retryCount;
+
+            $this->logger->info('Error returned by the server. Retrying #{retryCount} using {delay} ms delay. "{error}"', [
```

Comment:
I would suggest to remove the double quotes around the error, add a colon instead of a dot before, and also inline its value instead of passing through the context, to improve logs.